### PR TITLE
Mipmapping for global illumination

### PR DESCRIPTION
**Pull request by VZout**
* Mipmapping for global illumination
* Better normal map calculation for rt.

### CODE REVIEW

File: resources/shaders/dxr_functions.hlsl [hunk 1]
```diff
@@ -33,4 +33,34 @@ float3 unpack_position(float2 uv, float depth, float4x4 inv_vp)
...
 }
 
-#endif //__DXR_FUNCTIONS_HLSL__
\ No newline at end of file
+float3 CalcPeturbedNormal(float3 normal, float3 normal_map, float3 tangent, float3 bitangent, float3 V, out float3 world_normal)
+{
+	float3x4 object_to_world = ObjectToWorld3x4();
+	float3 N = normalize(mul(object_to_world, float4(normal, 0)));
+	float3 T = normalize(mul(object_to_world, float4(tangent, 0)));
+#define CALC_B
```
Comment (tntmeijs) [inline resources/shaders/dxr_functions.hlsl]: Can this be moved to the top of the file so we can have all configuration properties in an easy-accessible location? Hiding it down here makes it a bit more difficult to control shader behavior.

File: resources/shaders/dxr_functions.hlsl [hunk 2]
```diff
@@ -33,4 +33,34 @@ float3 unpack_position(float2 uv, float depth, float4x4 inv_vp)
...
+#ifndef CALC_B
+	const float3 B = normalize(mul(object_to_world, float4(bitangent, 0)));
+#else
+	T = normalize(T - dot(T, N) * N);
+	float3 B = cross(N, T);
+#endif
+	const float3x3 TBN = float3x3(T, B, N);
+
+	//float3 fN = N;
```
Comment (tntmeijs) [inline resources/shaders/dxr_functions.hlsl]: Old code?

File: resources/shaders/dxr_functions.hlsl [hunk 3]
```diff
@@ -33,4 +33,34 @@ float3 unpack_position(float2 uv, float depth, float4x4 inv_vp)
...
+	float3 fN = normalize(mul(normal_map, TBN));
+	fN = lerp(fN, -fN, dot(fN, V) < 0);
+
+	world_normal = N;
+
+	return fN;
+}
+
+float3 CalcPeturbedNormal(float3 normal, float3 normal_map, float3 tangent, float3 bitangent, float3 V)
```
Comment (tntmeijs) [inline resources/shaders/dxr_functions.hlsl]: Why use temp here? Isn't it cleaner to just pass "0" immediately?

File: resources/shaders/dxr_pathtracer_functions.hlsl [hunk 4]
```diff
@@ -48,8 +77,10 @@ float probabilityToSampleDiffuse(float3 difColor, float3 specColor)
 	return lumDiffuse / (lumDiffuse + lumSpecular);
 }
 
+static RayCone null_cone = { 0, 0 };
+
 float3 ggxDirect(float3 hit_pos, float3 fN, float3 N, float3 V, float3 albedo, float metal, float roughness, unsigned int seed, unsigned int depth)
-{
+ {
```
Comment (LaisoEmilio) [inline resources/shaders/dxr_pathtracer_functions.hlsl]: O M G xD

File: resources/shaders/dxr_pathtracer_functions.hlsl [hunk 5]
```diff
@@ -34,6 +33,36 @@ float3 TraceColorRay(float3 origin, float3 direction, unsigned int depth, unsign
...
+	{
+		return float3(0, 0, 0);
+	}
+
+	// Define a ray, consisting of origin, direction, and the min-max distance values
+	RayDesc ray;
+	ray.Origin = origin;
+	ray.Direction = direction;
+	ray.TMin = 0;
+	ray.TMax = 1.0e38f;
```
Comment (tntmeijs) [inline resources/shaders/dxr_pathtracer_functions.hlsl]: Isn't there a way to get this value via HLSL of something like that? Just curious... I don't like "magic numbers", hence I'm asking.

File: resources/shaders/dxr_pathtracer_functions.hlsl [hunk 6]
```diff
@@ -136,9 +167,17 @@ float3 ggxIndirect(float3 hit_pos, float3 fN, float3 N, float3 V, float3 albedo,
...
+		if (use_raycone)
+		{
+			TraceColorRayCone(hit_pos + (EPSILON * N), rand_dir, depth, seed, cone);
+		}
+		else
+		{
+			TraceColorRay(hit_pos + (EPSILON * N), rand_dir, depth, seed);
+		}
 
+		if (dot(N, rand_dir) <= 0.0f) irradiance = float3(0, 0, 0);
```
Comment (tntmeijs) [inline resources/shaders/dxr_pathtracer_functions.hlsl]: Perhaps split this up into multiple lines? The statement right after the check could potentially be a place for future bugs due to the lack of "{" and "}".

File: resources/shaders/dxr_pathtracer_functions.hlsl [hunk 7]
```diff
@@ -181,11 +220,19 @@ float3 ggxIndirect(float3 hit_pos, float3 fN, float3 N, float3 V, float3 albedo,
...
+		float3 irradiance = 0;
+		if (use_raycone)
+		{
+			irradiance = TraceColorRayCone(hit_pos + (EPSILON * N), L, depth, seed, cone);
+		}
+		else
+		{
+			irradiance = TraceColorRay(hit_pos + (EPSILON * N), L, depth, seed);
+		}
 		if (dot(N, L) <= 0.0f) irradiance = float3(0, 0, 0);
```
Comment (tntmeijs) [inline resources/shaders/dxr_pathtracer_functions.hlsl]: Perhaps split this up into multiple lines? The statement right after the check could potentially be a place for future bugs due to the lack of "{" and "}".

File: resources/shaders/dxr_pathtracer_main.hlsl [hunk 8]
```diff
@@ -79,10 +80,19 @@ void RaygenEntry()
...
 	float3 result = float3(0, 0, 0);
 
+	SurfaceHit sfhit;
+	sfhit.pos = wpos;
+	sfhit.normal = normal;
+	sfhit.dist = length(cpos - wpos);
+	sfhit.surface_spread_angle = ComputeSurfaceSpreadAngle(gbuffer_depth, gbuffer_normal, inv_vp, wpos, normal);
+
+	// Compute the initial ray cone from the gbuffers.
+ 	RayCone cone = ComputeRayConeFromGBuffer(sfhit, 1.39626, DispatchRaysDimensions().y);
```
Comment (tntmeijs) [inline resources/shaders/dxr_pathtracer_main.hlsl]: I don't like the magic number.

File: resources/shaders/dxr_functions.hlsl (hunk 3 above)
Comment (VZout) [reply resources/shaders/dxr_functions.hlsl]: can't do that since it is an out value

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 5 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: not new code.

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 6 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: not new code.

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 7 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: not new code.

File: resources/shaders/dxr_pathtracer_main.hlsl (hunk 8 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_main.hlsl]: is already an issue and out of the scope of this pr.

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 5 above)
Comment (tntmeijs) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: So it's not worth resolving? If not, feel free to commit the file the way it is. :)

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 7 above)
Comment (tntmeijs) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: So it's not worth resolving? If not, feel free to commit the file the way it is. :)

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 6 above)
Comment (tntmeijs) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: So it's not worth resolving? If not, feel free to commit the file the way it is. :)

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 7 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: not new code.

File: resources/shaders/dxr_functions.hlsl (hunk 2 above)
Comment (VZout) [reply resources/shaders/dxr_functions.hlsl]: removed

File: resources/shaders/dxr_functions.hlsl (hunk 1 above)
Comment (VZout) [reply resources/shaders/dxr_functions.hlsl]: yep

File: resources/shaders/dxr_pathtracer_main.hlsl (hunk 8 above)
Comment (tntmeijs) [reply resources/shaders/dxr_pathtracer_main.hlsl]: Okay, no need to change it, then! :)

File: resources/shaders/dxr_pathtracer_functions.hlsl (hunk 4 above)
Comment (VZout) [reply resources/shaders/dxr_pathtracer_functions.hlsl]: fixed

**Pull request merged**